clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2206,10 +2228,18 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
     return true;
   }
 
-  bool leaveStruct(const CXXRecordDecl *, const CXXBaseSpecifier &,
+  bool leaveStruct(const CXXRecordDecl *RD, const CXXBaseSpecifier &B,
                    QualType) final {
     --StructBaseDepth;
-    return true;
+    // FreeFunctionLoc.isInvalid() shows if checker object was created for a
+    // free function
+    if (FreeFunctionLoc.isInvalid())
+      return true;
+    if (B.isVirtual())
+      SemaSYCLRef.SemaRef.Diag(FreeFunctionLoc,
+                               diag::err_sycl_kernel_virtual_arg)
+          << RD->getNameAsString() << B.getType().getAsString();
+    return isValid();
```

Comment:
The diagnostic should not be specific to free functions and we need to set to invalid to stop visiting, and there is a `Diag` object already.
```suggestion
    // FreeFunctionLoc.isInvalid() shows if checker object was created for a
    // free function. If that is the case, point to the free function declaration.
    if (B.isVirtual()) {
      Diag(FreeFunctionLoc.isInvalid()? RD->getLocation() : FreeFunctionLoc,
                               diag::err_sycl_kernel_virtual_arg)
          << RD->getNameAsString() << B.getType().getAsString();
          IsInvalid = true;
   }
    return isValid();
```